### PR TITLE
Reworked Scenarios and Scenario Modifiers, Tied Civilian Units to Campaign Era

### LINKED ISSUE #1416: Create Era-appropriate Civilian Mech RATs
### Environment
MekHQ 0.47.2
Win10
Java (build 1.8.0_241-b07)

### Description
An extraction mission in 3033 generated a Quasit MilitiaMech QUA-51P, which isn't available until 3074.  The MHQ options Limit Units and Part by Year, Variable Tech Level, and Use Faction Intro Dates are enabled.  Only the QUA-51T is available under Purchase Unit.  Plasma rifle are flagged as unavailable under Purchase Parts.

Quasit

### Files
[Steel Recon30331101-PlasmaRifle.zip](https://github.com/MegaMek/mekhq/files/4088790/Steel.Recon30331101-PlasmaRifle.zip)

[customs.zip](https://github.com/MegaMek/mekhq/files/4088793/customs.zip)

### LINKED ISSUE #4925: [0.50.01-SNAPSHOT] Unreasonable Clan Opfor from Generic BV Clan Forces (Dropship Assault)
### Prerequisites and Pre-Issue Checklist

- [X] I'm reporting the issue to the correct repository:
- [MegaMek](https://github.com/MegaMek/megamek/issues)
- [MegaMekLab](https://github.com/MegaMek/megameklab/issues)
- [MekHQ](https://github.com/MegaMek/mekhq/issues)

- [X] I've tested the issue against at least the latest MILESTONE version
- [X] I've asked on the MegaMek Discord about the error
- [X] I've reviewed the BattleTech rules and MekHQ documentation, and I've confirmed that something isn't working as intended.
- [X] I've searched the Github tracker and haven't found the issue listed

### Severity *

Medium (Gameplay Limitation): Non-core functionality is impaired, providing a suboptimal but playable experience.

### Brief Description *

image

The force picked as the seed for the opfor is about 12.5k BV.

Enemy "Primary Enemy Opfor" force is 167% above the seed force, so within acceptable parameters, but the addition of the pocket warship worth 25k BV makes this scenario utterly unwinnable.

Additionally, the opfor has a full star of artillery, spawned from an artillery base on-track that I have not dealt with yet.

### Steps to Reproduce

N/A

### Operating System *

Windows 11

### Java Version *

Java Version: 17.0.6

### MekHQ Suite Version *

Free Text (type manually)

### Custom MekHQ Version

megamek.MegaMek.initializeLogging(MegaMek.java:146) - Starting MegaMek v0.50.1-SNAPSHOT     Build Date: 2024-09-26T01:26:30.505226736     Today: 2024-09-27     Origin Project: MegaMekLab     Java Vendor: Eclipse Adoptium     Java Version: 17.0.6     Platform: Windows 10 10.0 (amd64)     System Locale: en_US     Total memory available to MegaMek: 8 GB     MM Code Revision: d41eb83a27e1d0a31c430bb1749e242394c71d01     MML Code Revision: 1b955537a117a5f05d20287d29a2d4f13117b01d     MHQ Code Revision: 621775698d20e6bb8b6ef8d55600a79af7d4593a

### Attach Files

[Opfor Generation.zip](https://github.com/user-attachments/files/17171134/Opfor.Generation.zip)

### Final Checklist

- [X] I've checked to make sure that this issue has not already been filed
- [X] I'm reporting only one issue in this ticket for clarity and focus

### PR DESCRIPTION
**Pull request by IllianiCBT**
This PR is one I've been wanting to work on for a while, but haven't found the time. It goes through and rebalances scenario objectives across the board, taking a more holistic view of scenario objectives to prevent some of the weird behavior and objective bloat we've seen over the years.

I went through and rejigged the two scenario objectives that weren't tracked by mhq: Recon and Extract. Now for 'recon' you need to conduct the battle for a certain amount of time against an overwhelming enemy force. To represent the time taken to scan the battlefield or some such. For Extract, if the buildings are intact at the end of the scenario you'll get a Support Point.

I also went through and rebalanced our battery of scenario modifiers. Only one was removed (Search and Rescue, as that cannot be tracked in mhq). My goal here was to better balance the modifiers between each other.

Finally, I renamed and re-jigged the descriptions for the various scenarios types to make them feel less sterile. I also added a handful of new scenarios. It's my hope that this will make things feel fresher, even if the actual scenarios themselves haven't changed that much.

-------

Moving away from scenarios and modifiers, and onto contract command levels.

I wanted to make playing under House, Independent, Integrated, and Liaison command levels to feel different.

**House:** You will deploy with an allied force roughly the same size as your force. This represents your employer assigning joint forces to ensure objectives are completed. The idea here is that your employer has employed you to supplement their forces, to achieve objectives neither you nor they could complete alone.

**Integrated:** You will deploy with a large allied force. This represents you being integrated into your employers' wider organization. You are a cog in their machine.

**Liaison:** Your employer assigns you a liaison, whose job is to monitor your activities to ensure your employers' goals are met. To ensure this happens they will deploy a forward observer to keep an eye on you in the field. That liaison will always be in a heavy or assault unit. Apparently the employers of the Inner Sphere have gotten fed up with their liaisons getting shot out of the Locusts they all seemed to love so much.

**Independent:** You're on your own. Good luck, Commander.

-------

I went in and used functionality added by @SuperStucco (back in #4418) to remove the old fixed Civilian tables. Now we take advantage of unit roles to generate convoys and civilian units based on unit type and role.

I also added some fallback values to better catch units that failed to generate. Generally this will mean the unit generates as a tank (for ground scenarios), or Civilian DropShip (for DropShip scenarios). So while the unit might not be the type expected, it should better avoid instances where we generate empty forces. It can't completely remove the chance, but it should at least reduce the likelihood. As we find more holes in the system we can add additional checks and balances. This isn't meant to completely remove the issue, just to mitigate it.

### Closes #1416 && #4925 && #4850 && #4877 && #4927

### CODE REVIEW

File: MekHQ/data/scenariomodifiers/FacilityAlliedDefend.xml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <AtBScenarioModifier>
-    <additionalBriefingText>Preserve 50% of the designated turrets.</additionalBriefingText>
+    <additionalBriefingText>Preserve 25% of the designated turrets.</additionalBriefingText>
```
Comment (IllianiCBT) [inline MekHQ/data/scenariomodifiers/FacilityAlliedDefend.xml:2]: This change was to hold base defenses to a similar standard as base assaults. Previously, players were required to keep 50% of the base turrets alive, while they were required to kill 75-100% to destroy an enemy base.

File: MekHQ/data/scenariomodifiers/HouseOfficerAir.xml [hunk 2]
```diff
@@ -15,77 +14,21 @@
...
+        <minWeightClass>2</minWeightClass>
+        <retreatThreshold>0</retreatThreshold>
         <startingAltitude>5</startingAltitude>
+        <objectiveLinkedForces>
+            <objectiveLinkedForce>Player</objectiveLinkedForce>
+        </objectiveLinkedForces>
         <syncDeploymentType>SameEdge</syncDeploymentType>
         <syncedForceName>Player</syncedForceName>
         <useArtillery>false</useArtillery>
     </forceDefinition>
```
Comment (IllianiCBT) [inline MekHQ/data/scenariomodifiers/HouseOfficerAir.xml:33]: We're removing the various House Officer, Liaison, and Integrated Allies objectives and instead just having them hook into whatever the players' force objective is.

File: MekHQ/data/scenariomodifiers/SearchAndRescue.xml
Comment (IllianiCBT) [inline MekHQ/data/scenariomodifiers/SearchAndRescue.xml:1]: You'll see a lot of 'This file was deleted' this is the sole instance where I deleted something, the rest were all renamed.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 3]
```diff
@@ -2464,6 +2470,34 @@
...
+            // Roll for unit types
+            if (totalWeight <= 0) {
+                actualUnitType = UnitType.TANK;
+            } else {
+                int roll = Compute.randomInt(totalWeight);
+                if (roll < vehicleLanceWeight) {
+                    actualUnitType = UnitType.TANK;
+                // Mixed units randomly select between Mek or ground vehicle
+                } else {
+                    for (int x = 0; x < unitCount; x++) {
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java:2490]: No it isn't

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 4]
```diff
@@ -2867,6 +2974,35 @@
...
+     * @return a new instance of Entity with the specified parameters
+     */
+    private static Entity getNewEntity(String faction, SkillLevel skill, int quality,
+                                       List<Integer> unitTypes, String weights,
+                                       @Nullable Map<Integer, Collection<MissionRole>> rolesByType,
+                                       Campaign campaign, int i) {
+        Collection<MissionRole> roles;
+
+        if (rolesByType != null) {
+            roles = rolesByType.getOrDefault(unitTypes.get(i), new ArrayList<>());
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java:2998]: It's literally checked the line above, silly bot

**Comment by IllianiCBT** [comment]
Flipping this to live. Description has been updated accordingly.

**Review by HammerGS — APPROVED** [review]
Approved.

**Pull request merged**